File: opensource/kernel/src/main/java/com/sos/scheduler/engine/kernel/order/OrderSubsystem.java
package com.sos.scheduler.engine.kernel.order;

import com.google.common.base.Predicate;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Iterables;
import com.sos.scheduler.engine.data.folder.JobChainPath;
import com.sos.scheduler.engine.data.order.OrderKey;
import com.sos.scheduler.engine.kernel.cppproxy.Order_subsystemC;
import com.sos.scheduler.engine.kernel.job.Job;
import com.sos.scheduler.engine.kernel.order.jobchain.JobChain;
import com.sos.scheduler.engine.kernel.scheduler.SchedulerException;
import com.sos.scheduler.engine.kernel.scheduler.Subsystem;

import javax.annotation.Nullable;

public class OrderSubsystem implements Subsystem {
    private final Order_subsystemC cppProxy;

    public OrderSubsystem(Order_subsystemC cppproxy) {
        this.cppProxy = cppproxy;
    }

    public final ImmutableList<JobChain> jobChains() {
        return ImmutableList.copyOf(cppProxy.java_file_baseds());
    }

    public final Iterable<JobChain> jobchainsOfJob(final Job job) {
        return Iterables.filter(jobChains(), new Predicate<JobChain>() {
            @Override public boolean apply(JobChain jobChain) {
                return jobChain.refersToJob(job);
            }
        });
    }

    public final JobChain jobChain(JobChainPath o) {
        JobChain result = cppProxy.java_file_based_or_null(o.asString());
        if (result == null)
            throw new SchedulerException("Unknown '"+o+"'");
        return result;
    }

    @Nullable public final Order orderOrNull(OrderKey orderKey) {
        return jobChain(orderKey.jobChainPath()).orderOrNull(orderKey.id());
    }

    public final Order order(OrderKey orderKey) {
        return jobChain(orderKey.jobChainPath()).order(orderKey.id());
    }

    public  final void tryRemoveOrder(OrderKey k) {
        Order o = orderOrNull(k);
        if (o != null)
            o.remove();
    }
}

File: opensource/kernel/src/main/scala/com/sos/scheduler/engine/kernel/order/OrderSubsystem.scala
package com.sos.scheduler.engine.kernel.order

import com.google.inject.Injector
import com.sos.scheduler.engine.common.inject.GuiceImplicits._
import com.sos.scheduler.engine.data.filebased.FileBasedType
import com.sos.scheduler.engine.data.jobchain.JobChainPath
import com.sos.scheduler.engine.data.order.OrderKey
import com.sos.scheduler.engine.data.scheduler.ClusterMemberId
import com.sos.scheduler.engine.kernel.async.SchedulerThreadCallQueue
import com.sos.scheduler.engine.kernel.async.SchedulerThreadFutures.inSchedulerThread
import com.sos.scheduler.engine.kernel.cppproxy.{Job_chainC, Order_subsystemC}
import com.sos.scheduler.engine.kernel.filebased.FileBasedSubsystem
import com.sos.scheduler.engine.kernel.job.Job
import com.sos.scheduler.engine.kernel.order.jobchain.JobChain
import com.sos.scheduler.engine.kernel.persistence.hibernate._
import javax.inject.{Singleton, Inject}
import javax.persistence.EntityManagerFactory

@Singleton
final class OrderSubsystem @Inject private(
  protected[this] val cppProxy: Order_subsystemC, injector: Injector,
  implicit val schedulerThreadCallQueue: SchedulerThreadCallQueue)
extends FileBasedSubsystem {

  type MySubsystem = OrderSubsystem
  type MyFileBased = JobChain
  type MyFile_basedC = Job_chainC

  val companion = OrderSubsystem

  private[order] lazy val clusterMemberId = injector.apply[ClusterMemberId]
  private[order] lazy val entityManagerFactory = injector.apply[EntityManagerFactory]
  private[order] lazy val jobChainStore = injector.apply[HibernateJobChainStore]
  private[order] lazy val jobChainNodeStore = injector.apply[HibernateJobChainNodeStore]

//  def jobChainMap = new Map[JobChainPath, JobChain] {
//    def get(key: JobChainPath) =
//      jobChainOption(key)
//
//    def iterator =
//      jobChains.iterator map { o => o.path -> o }
//
//    def -(key: JobChainPath) =
//      throw new NotImplementedError
//
//    def +[B1 >: JobChain](kv: (JobChainPath, B1)) =
//      throw new NotImplementedError
//  }

  def tryRemoveOrder(k: OrderKey) {
    inSchedulerThread {
      for (o <- orderOption(k))
        o.remove()
    }
  }

  def order(orderKey: OrderKey): Order =
    jobChain(orderKey.jobChainPath).order(orderKey.id)

  def orderOption(orderKey: OrderKey): Option[Order] =
    jobChain(orderKey.jobChainPath).orderOption(orderKey.id)

  def removeJobChain(o: JobChainPath) {
    jobChain(o).remove()
  }

  def jobChainsOfJob(job: Job): Iterable[JobChain] =
    inSchedulerThread {
      jobChains filter { _ refersToJob job }
    }

  def jobChains: Seq[JobChain] =
    fileBaseds

  def jobChain(o: JobChainPath): JobChain =
    fileBased(o)

  def jobChainOption(o: JobChainPath): Option[JobChain] =
    fileBasedOption(o)
}


object OrderSubsystem extends FileBasedSubsystem.Companion[OrderSubsystem, JobChainPath, JobChain](FileBasedType.jobChain, JobChainPath.apply)
